### PR TITLE
Refactor hf transformers to be `TorchImageModels`

### PR DESCRIPTION
**Pull request by jacobsela**
## What changes are proposed in this pull request?

draft for now. Info in Jira.

## How is this patch tested? If it is not, please explain why.

(Details)

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
  - Enhanced support for Hugging Face Transformers models, including flexible configuration options and improved integration for image classification, object detection, semantic segmentation, and depth estimation tasks.
  - Added advanced embeddings extraction, aggregation methods, and dynamic handling of zero-shot prompts and class-specific prompts.
  - Introduced new output processors for seamless conversion of model outputs into FiftyOne label formats.

- **Refactor**
  - Unified and streamlined the integration of Hugging Face models, providing a more consistent and configurable user experience.

### CODE REVIEW

File: fiftyone/utils/transformers.py [hunk 1]
```diff
@@ -1029,62 +1144,192 @@ def _get_base_model_name(model):
...
+def _get_image_size(img):
+    if isinstance(img, torch.Tensor):
+        height, width = img.size()[-2:]
+    elif isinstance(img, Image.Image):
+        width, height = img.size
+    elif isinstance(img, np.ndarray):
+        height, width = img.shape[:2]
+
+    return height, width
+
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:1297]: not doing this for consistent with `TorchImageModel`. Should be refactored anyways.

File: fiftyone/utils/transformers.py [hunk 2]
```diff
@@ -339,56 +349,155 @@ class FiftyOneZeroShotTransformerConfig(FiftyOneTransformerConfig):
...
+                return key
+
+        else:
+            # have to set this to true to get the hidden states
+            self._output_hidden_states = True
+            return "hidden_states"
 
     def embed(self, arg):
-        return self._embed(arg)[0]
+        return self.embed_all([arg])[0]
```
Comment (manushreegangwar) [inline fiftyone/utils/transformers.py:471]: Is empty list return value possible here? If so, add a check.

**Review by manushreegangwar — CHANGES_REQUESTED** [review]
Few comments but looks good otherwise.

File: fiftyone/core/models.py [hunk 3]
```diff
@@ -745,7 +745,11 @@ def collate_fn(batch):
                 return error
 
             try:
-                return tud.dataloader.default_collate(batch)
+                return (
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/core/models.py:748]: Can't unfortunately, it's blocking. Will need to merge #5834 before this.

File: fiftyone/utils/transformers.py [hunk 4]
```diff
@@ -312,23 +246,99 @@ def _convert_bounding_box(box, image_shape):
...
+            self.name_or_path
+        )
+
+        # load classes if they exist
+        if self.hf_config.id2label is not None:
+            if self.classes is None:
+                # only load classes if they are not already set
+                self.classes = [
+                    self.hf_config.id2label[i]
+                    for i in range(len(self.hf_config.id2label))
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:307]: I'm doing it this way to ensure order. `values` doesn't guarantee order in a regular `dict` which [id2label is](https://huggingface.co/docs/transformers/v4.51.3/en/main_classes/configuration#transformers.PretrainedConfig.id2label).

File: fiftyone/utils/transformers.py [hunk 5]
```diff
@@ -7,19 +7,24 @@
...
+from fiftyone.core.models import EmbeddingsMixin, PromptMixin
 from fiftyone.zoo.models import HasZooModel
+import fiftyone.utils.torch as fout
 
-torch = fou.lazy_import("torch")
-transformers = fou.lazy_import("transformers")
+fou.ensure_torch()
+import torch
+
+fou.ensure_package("transformers")
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:26]: changed

File: fiftyone/zoo/models/manifest-torch.json [hunk 6]
```diff
@@ -2808,6 +2879,74 @@
...
+                "transformers",
+                "zero-shot"
+            ],
+            "date_added": "2024-01-17 14:25:51"
+        },
+        {
+            "base_name": "omdet-turbo-swin-tiny-torch",
+            "base_filename": null,
+            "version": null,
+            "description": "Hugging Face Transformers OWL-ViT",
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/zoo/models/manifest-torch.json:2920]: fixed

File: fiftyone/utils/transformers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:307]: I could do a sorted keys thing and iterate over that since they are guaranteed to be ints, but I think if a model has ints missing, it's on the model 🤷

File: fiftyone/utils/transformers.py [hunk 7]
```diff
@@ -312,23 +246,99 @@ def _convert_bounding_box(box, image_shape):
...
+                    for i in range(len(self.hf_config.id2label))
+                ]
+            else:
+                # if they are, this is either a zero shot model
+                # and the AutoConfig set some strange classes
+                # or the user is trying to override the classes
+                # in the latter case, strange things can happen
+                # not sure how to warn the user about this
+                logger.warning(
+                    "The classes passed to the FiftyOne model are different from the "
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:316]: changed

File: fiftyone/utils/transformers.py [hunk 8]
```diff
@@ -339,56 +349,155 @@ class FiftyOneZeroShotTransformerConfig(FiftyOneTransformerConfig):
...
+        if self.last_output is None:
+            raise ValueError(
+                "No embeddings have been generated yet. Call predict() first."
+            )
 
-        _dynamic_import = __import__(
-            module_name, fromlist=[classif_model_name, detection_model_name]
-        )
+        if self.embeddings_output_key is None:
+            self.resolve_embeddings_output_key(self.last_output)
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:399]: good catch, fixed

File: fiftyone/utils/transformers.py (hunk 2 above)
Comment (jacobsela) [reply fiftyone/utils/transformers.py:471]: not possible, would error out way beforehand if embeddings returned nothing

File: fiftyone/utils/transformers.py [hunk 9]
```diff
@@ -749,58 +908,30 @@ class FiftyOneZeroShotTransformerForObjectDetection(
...
+            config.output_processor_cls = "fiftyone.utils.transformers.TransformersDetectorOutputProcessor"
+        super().__init__(config)
+        # have to do this after init so processor is loaded
+        # I think this is better than instantiating a second one
+        # or passing the entire model to the output processor
+        self._output_processor.processor = self.transforms.processor
+        # ew
+        self.transforms.return_image_sizes = True
+        self.transforms.text = self.text_prompts
+        self.transforms.text_per_image = True
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:926]: It's a long story. TL;DR the way zero shot detection in HFT v4.51.3 (latest) works is that prompts have to be passed per image. In order to handle this, given that I don't know the batch ahead of time, I set that variable to be `True` here and then dynamically handle creating the batch in the transforms + collate.

File: fiftyone/utils/transformers.py [hunk 10]
```diff
@@ -1029,62 +1132,191 @@ def _get_base_model_name(model):
...
+    "base-model": FiftyOneTransformer,
+    "image-classification": FiftyOneTransformerForImageClassification,
+    "object-detection": FiftyOneTransformerForObjectDetection,
+    "semantic-segmentation": FiftyOneTransformerForSemanticSegmentation,
+    "depth-estimation": FiftyOneTransformerForDepthEstimation,
+    "zero-shot-image-classification": FiftyOneZeroShotTransformerForImageClassification,
+    "zero-shot-object-detection": FiftyOneZeroShotTransformerForObjectDetection,
+}
+
+# foo
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/transformers.py:1322]: done hahah

File: fiftyone/utils/transformers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (manushreegangwar) [reply fiftyone/utils/transformers.py:306]: In that case, I would suggest raising an exception so it doesn't look like FO integration is failing and it's the model `id2label` that is incorrect.

**Review by manushreegangwar — APPROVED** [review]
LGTM :+1:

**Review by kaixi-wang — APPROVED** [review]
core change looks fine

**Comment by brimoor** [comment]
@jacobsela please note that a `release/v1.6.0` branch already exists, and I believe you are intending this functionality to be used in that release, so you'll need to do one of two things:
1. Update this PR by rebasing its commits onto the `release/v1.6.0` branch
2. Merge this PR into `develop` and then create a new PR where you cherry-pick these commits onto `release/v1.6.0`

File: fiftyone/utils/transformers.py [hunk 11]
```diff
@@ -7,19 +7,23 @@
...
+from fiftyone.core.models import EmbeddingsMixin, PromptMixin
 from fiftyone.zoo.models import HasZooModel
+import fiftyone.utils.torch as fout
 
-torch = fou.lazy_import("torch")
-transformers = fou.lazy_import("transformers")
+fou.ensure_torch()
+import torch
+
+import transformers
```
Comment (brimoor) [inline fiftyone/utils/transformers.py:26]: Please switch to one of these two options for importing `transformers`:

```py
# Option 1
transformers = fou.lazy_import("transformers")
```

```py
# Option 2
fou.ensure_package("transformers")
import transformers
```

Either of these two options raises a curated error message that makes it clear to the user that we anticipated this situation and are telling them that they must install `transformers` in order to use this feature.

Conversely, if we just let the generic `ImportError` happen, it could cause users to think that it is our fault that we didn't list `transformers` as a requirement of `fiftyone`.

File: fiftyone/core/models.py [hunk 12]
```diff
@@ -745,7 +745,11 @@ def collate_fn(batch):
...
+                    tud.dataloader.default_collate(batch)
+                    if not hasattr(model, "collate_fn")
```
Comment (brimoor) [inline fiftyone/core/models.py:750]: Looks like this PR needs to be rebased onto https://github.com/voxel51/fiftyone/pull/5834, as this isn't how the final implementation of `collate_fn` landed.

File: fiftyone/utils/transformers.py (hunk 11 above)
Comment (jacobsela) [reply fiftyone/utils/transformers.py:27]: added back.
